src/coreclr/gc/gc.cpp

```diff
@@ -10796,6 +10829,25 @@ void gc_heap::return_free_region (heap_segment* region)
 {
     clear_region_info (region);
 
+    bool should_delete_region_p = global_region_allocator.should_delete_region (get_region_start (region),
+                                                                                num_free_regions,
+                                                                                num_free_large_regions);
+
+    if (should_delete_region_p)
+    {
+        // we need to either decommit or clear the committed space before
+        // giving it back to the region allocator - choice for now is to preferably decommit
+        uint8_t* page_start = align_lower_page (heap_segment_mem (region));
+        size_t size = heap_segment_committed (region) - page_start;
+        bool decommit_succeeded = virtual_decommit (page_start, size, heap_segment_oh (region), heap_number);
+        if (!decommit_succeeded)
+        {
+            memclr (page_start, size);
```

Comment:
do we want to log something here if the decommit fails?

Reply:
Yes, we should - will do...